lib/elixir/lib/calendar/datetime.ex

```diff
@@ -998,6 +998,68 @@ defmodule DateTime do
     naive_diff + System.convert_time_unit(offset_diff, :second, unit)
   end
 
+  @doc """
+  Adds a specified amount of time to a `DateTime`.
+
+  Accepts an `integer` in any `unit` available from `t:System.time_unit/0`.
+  Negative values will move backwards in time.
+
+  Takes changes such as summer time/DST into account. This means that adding time
+  can cause the wall time to "go backwards" during "fall back" during autumn.
+  Adding just a few seconds to a datetime just before "spring forward" can cause wall
+  time to increase by more than an hour.
+
+  Fractional second precision stays the same in a similar way to `NaiveDateTime.add/2`.
+
+  ### Examples
+
+      iex> dt = DateTime.from_naive!(~N[2018-11-15 10:00:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, dt_added} = dt |> DateTime.add(3600, :second, FakeTimeZoneDatabase)
+      iex> dt_added
+      #DateTime<2018-11-15 11:00:00+01:00 CET Europe/Copenhagen>
+
+      iex> dt = DateTime.from_naive!(~N[2018-11-15 10:00:00], "Etc/UTC")
+      iex> {:ok, dt_added} = dt |> DateTime.add(3600, :second)
+      iex> dt_added
+      #DateTime<2018-11-15 11:00:00Z>
+
+  When adding 3 seconds just before "spring forward" we go from 1:59:59 to 3:00:02
+
+      iex> dt = DateTime.from_naive!(~N[2019-03-31 01:59:59.123], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, dt_added} = dt |> DateTime.add(3, :second, FakeTimeZoneDatabase)
+      iex> dt_added
+      #DateTime<2019-03-31 03:00:02.123+02:00 CEST Europe/Copenhagen>
+
+  """
+  @doc since: "1.8.0"
+  @spec add(t, integer, System.time_unit(), Calendar.get_time_zone_database()) ::
+          {:ok, t()} | {:error, atom}
+  def add(
+        %DateTime{
+          utc_offset: utc_offset,
+          std_offset: std_offset,
+          calendar: calendar,
+          microsecond: {_, precision}
+        } = datetime,
+        integer_to_add,
+        unit \\ :second,
+        time_zone_database \\ Calendar.get_time_zone_database()
+      ) do
+    ppd = System.convert_time_unit(86400, :second, unit)
+    total_offset = System.convert_time_unit(utc_offset + std_offset, :second, unit)
+
+    datetime
+    |> to_iso_days()
+    # Subtract total offset in order to get UTC and add the integer for the addition
+    |> Calendar.ISO.add_day_fraction_to_iso_days(integer_to_add - total_offset, ppd)
+    |> from_iso_days(
+      %{time_zone: "Etc/UTC", zone_abbr: "UTC", utc_offset: 0, std_offset: 0},
+      calendar,
+      precision
+    )
+    |> shift_zone(datetime.time_zone, time_zone_database)
```

Comment:
We should probably extract some common paths between `shift_zone` and this function since we already have iso days. The current implementation will in most cases (other than UTC) go back and forth between the struct and iso days.

Reply:
Yes I considered doing that. Making a private function that takes iso days (and the other required information) and does the shifting.